### ISSUE TITLE
Compiling to PDF?

### ISSUE THREAD

**Issue by johnfrombluff** [bitcoinbook/bitcoinbook#399]
Is there a magic asciidoc or a2x incantation for compiling, either to HTML or PDF? 

Running `asciidoc book.asciidoc` produces errors relating to missing files and some other stuff, and `a2x -fpdf -dbook book.asciidoc` fails completely because xmllint fails to validate, not recognising the <a> tag, for example. 

I'm running Fedora, if that's relevant.

Thanks again to Andreas and O'Reilly for making this awesome resource available!

**Comment by jkleinwaechter** [comment]
I have a similar issue for generating an epub. My apologies ofr not being familiar with the xmllint program. I'll try to wade through it but thought you should know none-the-less.
Ditto on the thank you to Adnreas and O'Reilly.

a2x -f epub ./book.asciidoc

a2x: ERROR: "xmllint" --nonet --noout --valid "./book.xml" returned non-zero exit status 4

**Comment by jkleinwaechter** [comment]
@johnfrombluff It appears that simply adding the --noxmllint option allows the book to be built (at least in epub fashion).

a2x -f epub --no-xmllint ./book.asciidoc

**Comment by nonsequitur** [comment]
Here's a script to build the book in the following formats:
- html
- standalone html with embedded images
- epub
- pdf

(Just comment out unneded formats).

The script has just a single dependency: [Nix](https://nixos.org/nix/), which you can install on Linux and MacOS with
```sh
curl https://nixos.org/nix/install | sh
```
All other dependencies are fetched automatically.

```bash
#!/usr/bin/env bash

outputPaths=$(nix-build --no-out-link - <<'EOF'
with (import <nixpkgs> {});
let
  src = fetchFromGitHub rec {
      rev = "6354bb99d3834303693658890426a7ec6d4c104e";
      sha256 = "1w0p30d5p1yswfna1dw06ii5vajlij98b4x25bh7qzljb60hmivz";
      owner = "bitcoinbook";
      repo = "bitcoinbook";
      name = "mastering-bitcoin-src";
  };

  buildScripts = {
    htmlWithImgsDir = ''
      dest=$out/mastering_bitcoin_htm
      mkdir $dest
      asciidoc -a toc -o $dest/mastering_bitcoin.htm -d book book.asciidoc
      cp -r $src/images $dest
    '';

    standaloneHtml = "asciidoc -a data-uri -a icons -a toc -a max-width=55em -o $out/mastering_bitcoin_standalone.htm -d book book.asciidoc";

    epub = ''
      a2x -f epub -d book --no-xmllint --destination-dir=$out book.asciidoc
      mv $out/book.epub $out/mastering_bitcoin.epub
    '';

    pdf =
     # When building pdfs, a2x can only write its output to the source directory.
     # To work around this, copy src, build and delete src.
     ''
      cp -r $src $out/src
      cd $out/src
      chmod -R '+rw' .
      a2x -f pdf --fop -d book -L book.asciidoc
      mv $out/src/book.pdf $out/mastering_bitcoin.pdf
      rm -r $out/src
    '';
  };

  formatsToBuild = with buildScripts; [
    htmlWithImgsDir
    standaloneHtml
    epub
    pdf
  ];

  buildBook = buildScript:
     runCommand "mastering-bitcoin" {
      inherit src;
      buildInputs = [ asciidoc-full-with-plugins ];
    }
    (''
      mkdir $out
      cd $src
     '' + buildScript);

  bookOutputs = runCommand "mastering-bitcoin-outputs" {
    outputPaths = map buildBook formatsToBuild;
  } ''
    for x in $outputPaths; do
      echo $x/* >> $out
    done
  '';
in
  bookOutputs
EOF
) && printf "\n\n" && echo "Output paths:" && cat $outputPaths && echo

```

**Comment by b-onc** [comment]
@nonsequitur I got `error: file ‘nixpkgs’ was not found in the Nix search path (add it using $NIX_PATH or -I), at (string):1:14`

**Comment by kostaz** [comment]
After many trials... I got this error:
```
./nix.sh
these derivations will be built:
  /nix/store/hyj5nj9j0wc4pxk9b3cmp36clmpcw2hr-mastering-bitcoin.drv
  /nix/store/lp4q71bhnf83c7zcd6rv2p9bp21pf770-mastering-bitcoin.drv
  /nix/store/wfmbbngj02pp99rdc3qiq1zpdw7bvpdg-mastering-bitcoin.drv
  /nix/store/yw4dx9kk82hviw8gll6qrzccnw1n4nad-mastering-bitcoin.drv
  /nix/store/rqwnl52bxaxwmwbv04zihpkp97020sx5-mastering-bitcoin-outputs.drv
building path(s) ‘/nix/store/md3fnv6m1v7686wpgc5fizz1d7cipxf3-mastering-bitcoin’
a2x: ERROR: "/nix/store/9r7vjyhk2k43j4w7jv1nfydpr61xljsv-asciidoc-8.6.9/bin/asciidoc" --backend docbook -a "a2x-format=pdf"  --doctype book  --out-file "/nix/store/md3fnv6m1v7686wpgc5fizz1d7cipxf3-mastering-bitcoin/src/book.xml" "/nix/store/md3fnv6m1v7686wpgc5fizz1d7cipxf3-mastering-bitcoin/src/book.asciidoc" returned non-zero exit status 1
builder for ‘/nix/store/hyj5nj9j0wc4pxk9b3cmp36clmpcw2hr-mastering-bitcoin.drv’ failed with exit code 1
cannot build derivation ‘/nix/store/rqwnl52bxaxwmwbv04zihpkp97020sx5-mastering-bitcoin-outputs.drv’: 1 dependencies couldn't be built
error: build of ‘/nix/store/rqwnl52bxaxwmwbv04zihpkp97020sx5-mastering-bitcoin-outputs.drv’ failed
```

**Comment by nonsequitur** [comment]
@b-onc Sorry that this doesn't seem to work on macOS. I'll try if I can fix this later on my Mac.

@kostaz Try this updated [version of the script](https://gist.github.com/nonsequitur/23893618120ee24d8c2f0e61c39dfd72), where the build dependencies are fixed

**Comment by kostaz** [comment]
@nonsequitur ...
Hey, I tried it on Ubuntu 16.04 64-bit just a moment before...
```
➜  /tmp 
./build-mastering-bitcoin.sh  
error: attribute ‘fetchNixPkgs’ missing, at (string):14:14
```
Can you take a look?

**Comment by nonsequitur** [comment]
It's fixed.

**Comment by kostaz** [comment]
OMG! It worked!
http://docdro.id/6ErwaSm

**Comment by kingozymandius** [comment]
I couldn't get any links to the pdf to work. How annoying, why didn't he just release the pdf. This is just his way of saying it is an 'open source book' but still making you buy it...

**Comment by JoKeyser** [comment]
> @kingozymandius:  
> How annoying, why didn't he just release the pdf. This is just his way of saying it is an 'open source book' but still making you buy it...

I think your use of quotation marks around 'open source book' is wrong or you don't understand what it means. The book's source is clearly open. However that does not mean anyone has to make it more convenient for you, or print it for you, or read it aloud for you.
Not offering the PDF has two desirable effects in my opinion: A) Increased chance for compensation of the authors work through books sells, because some people evidently will prefer convenience over doing some research how to compile it. B) It encourages people to read the discussion and improvements about this book here on GitHub, where the sausage is made. I prefer B), to see the latest version, and potentially contribute myself.

**Comment by hamzaali00001** [comment]
@nonsequitur 

> @b-onc Sorry that this doesn't seem to work on macOS. I'll try if I can fix this later on my Mac.
By the way: you don't have to use sudo when running nix.
> @kostaz Try this updated version of the script, where the build dependencies are fixed.

Thanks. It works.

**Comment by aantonop** [comment]
The 2nd edition of "Mastering Bitcoin" is available under a CC-BY-NC-ND license, not a CC-BY-SA license. 

It is deliberately not available as a PDF. Why? Because a PDF is a "derivative" product, which is what the ND prohibits. That's because the publisher (O'Reilly Media) is a for-profit publisher and puts considerable resources behind distributing the book. 

When you make a PDF, you are violating the license. Worse, many will simply distribute the PDF on torrents and publishing sites. Since this issue was posted, I've seen a significant increase in pirate sites offering the PDF to the second edition for free or undercutting the publisher's price. This reduces the incentives of publishers like O'Reilly to consider open licenses. 

The book will eventually (within a year of publication) be released under a CC-BY-SA license, at which point PDF and translations will be allowed. Until then, making PDF copies violates the license and hurts the publisher (and the author) ability to make a living on the considerable amount of work that went into this book. Furthermore, if you make it so the publisher can't recoup their investment, they will delay the release into CC-BY-SA.

tl;dr: The book is licensed so you can read it for free. It is not licensed so you can make a PDF of it to read it *conveniently*. Convenience costs money. Violating the license is a shitty thing to do, especially when this is one of the very few books that is licensed so openly. 

--- LEAVE THIS ISSUE OPEN ---

**Comment by alitchfi** [comment]
Thank you for the clarification from Jan 28.

Just so that I am clear, when running asciidoctor-pdf --trace book.asciidoc

and then getting

/Library/Ruby/Gems/2.3.0/gems/asciidoctor-pdf-1.5.0.alpha.16/lib/asciidoctor-pdf/index_catalog.rb:37:in `ord': empty string (ArgumentError)

Is deliberate?

I fixed all the other errors in formatting in the source.

**Comment by 0bs72** [comment]
I have a link to a PDF version, if it's of interest or permissible.

**Comment by aantonop** [comment]
The PDF ebook, is offered for sale on my shop at aantonop.com, together
with EPUB and MOBI formats, properly formatted.
